tonic/routes.go

```diff
@@ -0,0 +1,161 @@
+
+	router.PathPrefix("/assets/").Handler(http.StripPrefix("/assets/", http.FileServer(http.Dir("./assets"))))
+	return nil
+}
+
+func (srv *Tonic) renderLoginPage(w http.ResponseWriter, r *http.Request) {
+	tmpl := template.New("layout")
+	tmpl, err := tmpl.Parse(templates.Layout)
+	checkError(err)
+	tmpl, err = tmpl.Parse(templates.Login)
+	checkError(err)
+	tmpl.Execute(w, nil)
+}
+
+func (srv *Tonic) userLoginPost(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	username := r.FormValue("username")
+	password := r.FormValue("password")
+	if username == "" || password == "" {
+		srv.web.ErrorResponse(w, http.StatusUnauthorized, "authentication failed")
+		return
+	}
+
+	client := gogs.NewClient(srv.config.GINServer, "")
+	var userToken string
+	tokens, err := client.ListAccessTokens(username, password)
+	if err != nil {
+		srv.web.ErrorResponse(w, http.StatusUnauthorized, "authentication failed")
+		return
+	}
+
+	if len(tokens) == 0 {
+		token, err := client.CreateAccessToken(username, password, gogs.CreateAccessTokenOption{Name: "testbot"})
+		userToken = token.Sha1
+		if err != nil {
+			srv.web.ErrorResponse(w, http.StatusUnauthorized, "authentication failed")
+			return
+		}
+	} else {
+		userToken = tokens[0].Sha1
+	}
+
+	client = gogs.NewClient(srv.config.GINServer, userToken)
+	orgs, err := client.ListMyOrgs()
+
+	for idx, org := range orgs {
+		// Bot needs to be owner of org
+		fmt.Printf("%d: %s\n", idx, org.UserName)
+	}
+
+	// TODO: Session cookie with token in DB
+	cookie := http.Cookie{
+		Name:    srv.config.CookieName,
+		Value:   userToken, // create session IDs linked to token instead
+		Expires: time.Now().Add(7 * 24 * time.Hour),
+		Secure:  false,
+	}
+
+	http.SetCookie(w, &cookie)
+	// Redirect to form
+	http.Redirect(w, r, "/", http.StatusFound)
+}
+
+func (srv *Tonic) renderForm(w http.ResponseWriter, r *http.Request) {
+	tmpl := template.New("layout")
+	tmpl, err := tmpl.Parse(templates.Layout)
+	checkError(err)
+	tmpl, err = tmpl.Parse(templates.Form)
+
+	checkError(err)
+
+	elements := make([]Element, len(srv.form))
+	copy(elements, srv.form)
+
+	data := make(map[string]interface{})
+	data["elements"] = elements
+
+	if err := tmpl.Execute(w, data); err != nil {
+		log.Printf("Failed to render form: %v", err)
+	}
+}
+
+func (srv *Tonic) showJob(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	jobid, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		srv.web.ErrorResponse(w, http.StatusInternalServerError, "Invalid ID")
+	}
+	job, err := srv.db.GetJob(jobid)
+
+	if err != nil || job == nil {
+		srv.web.ErrorResponse(w, http.StatusNotFound, "No such job")
+		return
+	}
+
+	tmpl := template.New("layout")
+	tmpl, err = tmpl.Parse(templates.Layout)
+	checkError(err)
+	tmpl, err = tmpl.Parse(templates.Form)
+	checkError(err)
+
+	data := make(map[string]interface{})
+	elements := make([]Element, len(srv.form))
+	copy(elements, srv.form)
```

Comment:
Could not figure out how or where `srv.form` is actually set. `tonic` has a `SetForm` method but I could not find it being used anywhere.

Reply:
Each instance that uses the library is responsible for defining its own form.  Basically, in its current state, the library is supposed to make it easy to define a form and a function that uses the data from that form.  The idea is that with those two elements you can build a bunch of different services that ask users for arbitrary input and automate actions on GIN on behalf of the user.